ezmlm-store.1
.TH ezmlm-store 1
.SH NAME
ezmlm-store \- store messages posted to a moderated lists and send
moderation requests to the moderator(s).
.SH SYNOPSIS
.B ezmlm-store [-bBcCmMpPrRsSvVyY]
.I dir
.SH DESCRIPTION
.B ezmlm-store
stores a main message sent to a ezmlm mailing list awaiting
message moderation or user confirmation.

If the
.B \-y
option is set, or
.I dir\fB/confirmpost
exists and the
.B \-Y
option is not set, it works in user confirmation mode.  Otherwise if
.I dir\fB/modpost
exists, it works in message moderation mode.
Otherwise the message is passed on to
.BR ezmlm-send(1) .

If either
.I dir\fB/modpostonly
exists or the
.B \-P
option is set, only moderators will be allowed to post.  Posts from
other addresses are bounced.
.SH "USER CONFIRMATION"
.B ezmlm-store
reads a mail message and
stores it in
.IR dir\fB/mod/unconfirmed/ .
.B ezmlm-store
also sends a user confirmation request with a copy of the message to the
envelope sender of the message.
The confirm request includes cookies for message rejection and acceptance.

The sender can take either action by replying to the appropriate address.
For user convenience, the
.I confirm
address is given in
the ``Reply-To:'' header.

Once the confirmation request has been successfully sent,
.B ezmlm-store
sets the owner-execute bit of the message in
.IR dir\fB/mod/unconfirmed/ .
Messages without this bit set are the result of incomplete executions of
.B ezmlm-store
and are ignored.

.SH MODERATION
.B ezmlm-store
reads a mail message and
stores it in
.I dir\fB/mod/pending/.
.B ezmlm-store
also sends a moderation request with a copy of the message to the list of
moderators stored in
.I dir\fB/mod/subscribers/.
If
.I dir\fB/modpost
contains a directory name,
.IR moddir ,
with a leading forward slash, the moderator list in
.I /moddir/subscribers/
is used instead. This allows separate sets of moderators for
subscription and message moderation, as
well as the use of the same moderator list for several
.B ezmlm
mailing lists, owned by the same user. If the
.B \-s
switch is used, the moderators are divided into two sets. Those pointed to by
.I dir\fB/modpost
which will receive confirmation requests for posts with their SENDER address,
and those stored with the base directory
.I dir\fB/mod/
which will receive moderation requests for messages posted from SENDERs not
part of the first group. This can be used to have a set of trusted
users (``moderating'' themselves to thwart SENDER faking), and a set of more
traditional moderators, which receive approval requests for other posts.

.B Note:
The contents of
.I dir\fB/modpost
do not affect the location of the message moderation queue.

If SENDER is a moderator, the request is sent
to that moderator only.

The moderation request includes cookies for message rejection and acceptance.
The moderator can take either action by replying to the appropriate address.
For moderator convenience, the
.I accept
address is given in
the ``Reply-To:'' header, and the
.I reject
address in the ``From:'' header.

If the list is set up with remote administration and the sender is not a
moderator, the ``Cc:'' header will contain a confirmation address to add
the sender to the ``allow'' database. This allows the moderator to both
approve the post and add the sender to the list of accepted posters by
replying to ``all''.

The default subject of all message moderation requests is
.B MODERATE for
.IR listname@host .
The default subject of all message confirmation requests is
.B Confirm post to
.IR listname@host .

Once the request has been successfully sent to the moderators, 
.B ezmlm-store
sets the owner-execute bit of the message in
.I dir\fB/mod/pending/.
Messages without this bit set are the result of incomplete executions of
.B ezmlm-store
and are ignored.

.SH "CONFIRMATION AND MODERATION MESSAGES"
At the beginning of the message,
.B ezmlm-store
prints a new
.B Mailing-List
field with the contents of the
.I TXT_MAILING_LIST
message.  It rejects the message if there is already a
.B Mailing-List
field.

.B ezmlm-store
does not distribute bounce messages:
if the environment variable
.B SENDER
is set, and is either empty or
.BR #@[] ,
.B ezmlm-store
rejects the message.

If
.I dir\fB/qmqpservers
exists,
.B ezmlm-store will use
.B qmail-qmqp(1)
to send messages.
.SH OPTIONS
.TP
.B \-b
(Default.)
With the moderation request,
send both headers and body of the message to be moderated.
.TP
.B \-B
With the moderation request,
send header, but not body of the message to be moderated.
This may be useful if moderation is used to allow one or a few poorly
connected addresses to post, and posts are very large (e.g. when the list
is used as a distribution channel of e.g. catalogs).
.TP
.B \-cCrR
Passed on to
.B ezmlm-send(1)
in case this program is executed. This happens only if neither
.I dir\fB/modpost
nor
.I dir\fB/confirmpost
exist.
.TP
.B \-m
(Default.)
The post is sent as a MIME enclosure.
.TP
.B \-M
The post is appended to the message.
.TP
.B \-s
The contents of
.I dir\fB/modpost
with the default
.I dir\fB/mod/
are used as the base directory for self-approving users, whereas
.I dir\fB/mod/
is used as the base directory
for the moderators to whom approval requests for other
posts are sent.
.TP
.B \-S
(Default.)
The contents of
.I dir\fB/modpost
with the default
.I dir\fB/mod/
are used as the base directory
for the moderators for all moderator functions.
.TP
.B \-p
(Default.)
Anyone can post. Posts that are not from a moderator are sent out to
all moderators for approval. Posts that are from a moderator are
sent only to that moderator. This is the default unless the
.I dir\fB/modpostonly
file exists.
.TP
.B \-P
Only moderators can post. Those posts are sent to the posting moderator
for approval. Posts from other addresses are bounced. This can be used
to relatively securely restrict posts to messages originating from
a given set of addresses.
.TP
.B \-y
Run in post confirmation mode even if
.IR dir\fB/confirmpost
does not exists.
.TP
.B \-Y
Ignore
.IR dir\fB/confirmpost.
(This option is useful if
.B ezmlm-store
is called after post confirmation, to allow message moderation.)
.TP
.B \-v
Display
.B ezmlm-store
version information.
.TP
.B \-V
Display
.B ezmlm-store
version information.
.SH USAGE
First of all, remember that
.B ezmlm-store
will simply execute
.B ezmlm-send(1)
unless
.I dir\fB/modpost
exists,
.I dir\fB/confirmpost
exists,
or
.B \-y
option is specified.
For the switches, use
.B \-P
for lists where only moderators can post securely, i.e. they are
asked to approve their own posts after posting,
whereas other SENDERs are rejected. For
low-security more convenient versions of this type of restriction, see
.BR ezmlm-gate(1) .

Normally (no switches),
posts from moderators are sent for approval only to the sending
moderator, whereas posts from other addresses are sent to all moderators.
Thus, you can set up an announcement list by adding all potential posters
as moderators. However, you may also want to allow others to post, as
long as the posts are approved. This would be default
.B ezmlm-store
function.

To make the set of ``approved posters'' who approve their own posts, different
from the set of addresses that approve posts from other users, use the
.B \-s
switch. Add a directory name to
.IR dir\fB/modpost .
This directory is the base directory of the ``approved posters''
database. Add the moderators for other posts to
.I dir\fB/mod/
using
.BR ezmlm-sub(1) .
.SH "CHARACTER SETS"
If
.I dir\fB/charset
exists,
.B ezmlm-store
will use the character set listed for all messages. Otherwise, the
default ``us-ascii'' will be used. The character set can be suffixed
by ``:'' followed by a code. If the code is ``Q'', outgoing messages are 
sent as ``Quoted-Printable'', if it is '`B'' they are sent ``base64'' encoded.
Otherwise, text is sent as is.
.SH "SEE ALSO"
ezmlm-clean(1),
ezmlm-confirm(1),
ezmlm-manage(1),
ezmlm-make(1),
ezmlm-moderate(1),
ezmlm(5)


CHANGES
BUGS:	-ezmlm-idx does not implement a full rfc822 parser. Comments can be
	 inserted to defeat the MIME parsing mechanism, such as:
	 Content-Type: (comment) test/plain
	 This type of comment, while legal, does not occur in normal messages.
	 Guiding info past ezmlm-idx MIME parsing is easier done by enclosing
	 a message within a message. The mechanism isn't intended as a
	 "secure" way of preventing e.g. "application/ms-tnef" from being
	 disseminated.
	 If you think this is a problem, E-mail ezmlm@list.cr.yp.to.
	 Mimeremove/reject work only at the outer MIME layer.
	-ezmlm-request requires addresses to be user@host. While rfc822
	 compliant, addresses like ``"the man" @ host'' are not accepted.
	-ezmlm-cgi doesn't deal optimally with holes in the archive.
	-ezmlm-cgi seems not to work with Chinese. The problem has been to
	 find someone willing/able to participate in debugging.

Some ezmlmrc are not updated and may not work right.  If you fix them,
please send them to ezmlm@list.cr.yp.to for inclusion in the next
release.

- Added <#C#>, <#T#>, and <#X#> substitutions for (un)subscribe
  confirmation emails, replaced with the hash code, time stamp, and
  action respectively.

- Added option to ezmlm-[un]sub to use a tag other than "manual" in Log.

- Added option to ezmlm-manage to show what response is being sent.

- Fixed behavior of ezmlm-manage -Q flag to match man page.

- Fixed ezmlm-import failing to flush output to the last message,
  and enhanced it to allow reading the mbox from stdin.
  Thanks to Tullio Andreatta.

ezmlm-idx-7.1.1, 2010-11-18
===========================

- Fixed bug in getln2 function that prevented error handling.

- Fixed bug in ezmlm-cron that prevented parsing of spaces.

- Fixed unclosed file descriptor on error path in ezmlm-cgi.

ezmlm-idx-7.1.0, 2010-11-03
===========================

- Added support for SQLite3 subscriber databases.
  Thanks to Mike Tedder.

- Added support for wildcard addresses in all address databases.

- Added support for qmqpservers to all ezmlm-idx programs.

- Added support for decoding sender addresses mangled with a BATV "btv1" tag.

- Added support for custom subject lines to ezmlm-get (for digests).

- Added a new "omitbottom" control file, equivalent to the -B option for
  both ezmlm-get and ezmlm-manage.

- Modified the (un)subscribe procedure to not require confirmation
  before telling the sender they were already on (or off) the list.

- Renamed to "setup" makefile target to the more standard "install".

- Modified ezmlm-split to exit silently if there is not "split" file,
  and added it back into the "manager" file to properly handle subscribe
  and unsubscribe requests destined for sublists.

- Made the ezmlm-reject -h (obey headerreject) option the default.

- Fixed ezmlm-send to obey the "addtrailer" setting instead of looking
  for a "text/trailer" file.

- Fixed handling of adding the trailer on old lists.

- Fixed adding the trailer to posts encoded with base64.

- Fixed handling Received: headers with the date stamp on a separate line.

- Fixed bug in command-line option processing which caused sender-
  confirmed posting to fail.

- Fixed handling BATV sender addresses with upper-case hex tags.
  Thanks to Kyle Wheeler for pointing this out.

- Adjusted filename in digest attachments to fix problem with Outlook.
  Thanks to Glen Stewart

ezmlm-idx-7.0.2, 2009-08-09
===========================

- Fixed selection of [un]subscribe subject line for moderated
  subscriptions.

- Clarified ezmlm-reject man page to note that Precedence: bulk is also
  rejected.

- Added a proper charset for ch_GB (Chinese GuoBiao), and added an
  explicit charset for the other languages that were missing them.

- Switched all ISO-8859-1 charsets to the more modern ISO-8859-15.

- Many small text updates and tweaks.

ezmlm-idx-7.0.1, 2009-07-09
===========================

- Added support for decoding sender addresses mangled by BATV prvs.

- Added missing ezmlm-checksub to the installed programs.

- Improved the TXT_BY message used in digests to work better with
  non-English languages.

- Fixed out-of-memory resulting in a permanent error.

- Fixed a buffer underflow bug in concatHDR.

- Fixed a bug in ezmlm-issubn.c that was causing it to scan the main
  list when it shouldn't.

- Fixed missing defaults in ezmlm-tstdig when digsize, digcount, or
  digtime were not created.

- Clarified the steps necessary when upgrading in UPGRADE.

ezmlm-idx-7.0.0, 2008-06-16
===========================

This version has three major architectural changes:

1. There is now support for internationalized messages.  All error
messages, all subject lines, and a few other messages are now configured
in a file named "text/messages" that is read at run time.  Messages in
this file will undergo header-style substitution before they are output.

Unlike other text files, all 3 of the files (DIR/text/messages,
/etc/ezmlm/LANG/text/messages, and /etc/ezmlm/default/text/messages) are
read, and only the first match (in the above order) is used for any
given message.  This allows for creation of partial files to override
just select messages.  If no match is found, the internal English text
is used.

2. The use of the "flags" file has been deprecated completely by
individual flag files accessed by the appropriate programs.  This, along
with corresponding changes to the ezmlmrc template, allows for lists to
be reconfigured without invoking ezmlm-make.  To upgrade a list to the
new setup, simply run "ezmlm-make -+ DIR".  Note that this will delete
the "config" and "flags" files, as they have been superceded by other
files.

3. Creation of subscriber tables has been moved out of ezmlm-mktab-* and
into ezmlm-make by adding some additional hooks to the subdb plugins.  A
ezmlm-rmtab program is also added to remove subscriber tables, also with
hooks in the subdb plugins.  The ezmlm-mktab-* programs have been
obsoleted by these changes and have been removed.


ezmlm-store.c
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include "stralloc.h"
#include "subfd.h"
#include "strerr.h"
#include "error.h"
#include "qmail.h"
#include "env.h"
#include "sender.h"
#include "lock.h"
#include "sig.h"
#include "open.h"
#include "getln.h"
#include "str.h"
#include "fmt.h"
#include "readwrite.h"
#include "wait.h"
#include "exit.h"
#include "substdio.h"
#include "getconf.h"
#include "datetime.h"
#include "now.h"
#include "cookie.h"
#include "getconfopt.h"
#include "messages.h"
#include "byte.h"
#include "case.h"
#include "quote.h"
#include "hdr.h"
#include "die.h"
#include "idx.h"
#include "copy.h"
#include "subdb.h"
#include "mime.h"
#include "wrap.h"
#include "config.h"
#include "auto_version.h"

int flagmime = MOD_MIME;	/* default is message as attachment */
int flagmodpostonly = -1;	/* default anyone can post */
				/* =1 for only moderators can */
int flagself = 0;		/* `modpost` mods approve own posts */
				/* but mod/ is used for moderators */
				/* of other posts. Def=no=0 */
int flagconfirm = -1;           /* if true, sender must approve its own posts */
int flagbody = 1;		/* body of message enclosed with mod request */
				/* 0 => headers only */

const char FATAL[] = "ezmlm-store: fatal: ";
const char USAGE[] =
"ezmlm-store: usage: ezmlm-store [-cCmMpPrRsSvV] dir";

static stralloc sendopt = {0};

static struct option options[] = {
  OPT_FLAG(flagbody,'b',1,0),
  OPT_FLAG(flagbody,'B',0,0),
  OPT_FLAG(flagmime,'m',1,0),
  OPT_FLAG(flagmime,'M',0,0),
  OPT_FLAG(flagmodpostonly,'p',0,0), /* anyone can post (still mod'd) */
  OPT_FLAG(flagmodpostonly,'P',1,"modpostonly"), /* only moderators can post */
  OPT_FLAG(flagself,'s',1,0),	     /* modpost and DIR/mod diff fxns */
  OPT_FLAG(flagself,'S',0,0),	     /* same fxn */
  OPT_FLAG(flagconfirm,'y',1,"confirmpost"), /* force post confirmation */
  OPT_FLAG(flagconfirm,'Y',0,0),     /* disable post confirmation */
  OPT_COPY_FLAG(sendopt,'c'),
  OPT_COPY_FLAG(sendopt,'C'),
  OPT_COPY_FLAG(sendopt,'r'),
  OPT_COPY_FLAG(sendopt,'R'),
  OPT_END
};

stralloc fnmsg = {0};

void die_msg(void) { strerr_die2sys(111,FATAL,MSG1(ERR_WRITE,fnmsg.s)); }

int fdmsg;
int fdmod;
int pid;
int match;

char strnum[FMT_ULONG];
char hash[COOKIE];
char boundary[COOKIE];
datetime_sec when;
struct stat st;

stralloc fnbase = {0};
stralloc line = {0};
stralloc mydtline = {0};
stralloc returnpath = {0};
stralloc accept = {0};
stralloc action = {0};
stralloc reject = {0};
stralloc quoted = {0};
stralloc moderators = {0};

struct qmail qq;

int subto(const char *s,unsigned int l)
{
  qmail_put(&qq,"T",1);
  qmail_put(&qq,s,l);
  qmail_put(&qq,"",1);
  return (int) l;
}

substdio ssin;
char inbuf[1024];

substdio ssmsg;
char msgbuf[1024];

substdio sstext;
char textbuf[512];

substdio sssub;
char subbuf[512];

void makeacthash(stralloc *act)
/* act is expected to be -reject-ddddd.ttttt or -accept-ddddd.ttttt and
 * has to be 0-terminated. */
/* The routine will add .hash@outhost to act. act will NOT be 0-terminated */
{
  int d;

  d = 2 + str_chr(act->s + 1,'-');
  cookie(hash,key.s,key.len,act->s + d,"","a");
  *(act->s + act->len - 1) = '.';	/* we put a '.' Bad, but works */
  if (!stralloc_catb(act,hash,COOKIE)) die_nomem();
  if (!stralloc_cats(act,"@")) die_nomem();
  if (!stralloc_cat(act,&outhost)) die_nomem();
  set_cphash(hash);
}

int main(int argc,char **argv)
{
  const char *dir;
  int fdlock;
  const char *sender;
  int match;
  int flaginheader;
  int flagmodpost;
  int flagremote;
  int ismod;
  stralloc mod = {0};
  const char *err;
  unsigned int i;

  (void) umask(022);
  sig_pipeignore();

  sender = get_sender();

  if (sender) {
    if (!*sender || str_equal(sender,"#@[]"))
      strerr_die2x(100,FATAL,MSG(ERR_BOUNCE));
  }

  if (!stralloc_copys(&sendopt,"-")) die_nomem();
  getconfopt(argc,argv,options,1,&dir);
  initsub(0);

  flagmodpost = getconf_line(&moderators,"modpost",0);
  flagremote = getconf_isset("remote");
  if (!flagmodpost && !flagconfirm) /* not msg-mod. Exec ezmlm-send */
    wrap_execbin("/ezmlm-send", &sendopt, dir);

  if (!moderators.len) {
    if (!stralloc_copys(&moderators,"mod")) die_nomem();
  }
  if (!stralloc_0(&moderators)) die_nomem();

  if (sender) {
    ismod = issub(moderators.s,sender,&mod);
    closesub();
				/* sender = moderator? */
  } else
    ismod = 0;

  if (!ismod && flagmodpostonly)
    strerr_die2x(100,FATAL,MSG(ERR_NO_POST));

  fdlock = lockfile("mod/lock");

  if (!stralloc_copys(&mydtline, flagconfirm
    ? "Delivered-To: confirm to "
    : "Delivered-To: moderator for ")) die_nomem();
  if (!stralloc_catb(&mydtline,outlocal.s,outlocal.len)) die_nomem();
  if (!stralloc_append(&mydtline,"@")) die_nomem();
  if (!stralloc_catb(&mydtline,outhost.s,outhost.len)) die_nomem();
  if (!stralloc_cats(&mydtline,"\n")) die_nomem();

  if (!stralloc_copys(&returnpath,"Return-Path: <")) die_nomem();
  if (sender) {
    if (!stralloc_cats(&returnpath,sender)) die_nomem();
    for (i = 14; i < returnpath.len;++i)
      if (returnpath.s[i] == '\n' || !returnpath.s[i] )
        returnpath.s[i] = '_';
		/* NUL and '\n' are bad, but we don't quote since this is */
		/* only for ezmlm-moderate, NOT for SMTP */
  }
  if (!stralloc_cats(&returnpath,">\n")) die_nomem();

 pid = getpid();		/* unique file name */
 for (i = 0;;++i)		/* got lock - nobody else can add files */
  {
   when = now();		/* when is also used later for date! */
   if (!stralloc_copys(&fnmsg, flagconfirm?"mod/unconfirmed/":"mod/pending/")) die_nomem();
   if (!stralloc_copyb(&fnbase,strnum,fmt_ulong(strnum,when))) die_nomem();
   if (!stralloc_append(&fnbase,".")) die_nomem();
   if (!stralloc_catb(&fnbase,strnum,fmt_ulong(strnum,pid))) die_nomem();
   if (!stralloc_cat(&fnmsg,&fnbase)) die_nomem();
   if (!stralloc_0(&fnmsg)) die_nomem();
   if (stat(fnmsg.s,&st) == -1) if (errno == error_noent) break;
   /* really should never get to this point */
   if (i == 2)
     strerr_die2x(111,FATAL,MSG(ERR_UNIQUE));
   sleep(2);
  }

  if (!stralloc_copys(&action,"-")) die_nomem();
  if (!stralloc_cats(&action,flagconfirm?ACTION_DISCARD:ACTION_REJECT)) die_nomem();
  if (!stralloc_cat(&action,&fnbase)) die_nomem();
  if (!stralloc_0(&action)) die_nomem();
  makeacthash(&action);
  if (!quote(&quoted,&outlocal)) die_nomem();
  if (!stralloc_copy(&reject,&quoted)) die_nomem();
  if (!stralloc_cat(&reject,&action)) die_nomem();
  if (!stralloc_0(&reject)) die_nomem();

  if (!stralloc_copys(&action,"-")) die_nomem();
  if (!stralloc_cats(&action,flagconfirm?ACTION_CONFIRM:ACTION_ACCEPT)) die_nomem();
  if (!stralloc_cat(&action,&fnbase)) die_nomem();
  if (!stralloc_0(&action)) die_nomem();
  makeacthash(&action);
  if (!stralloc_copy(&accept,&quoted)) die_nomem();
  if (!stralloc_cat(&accept,&action)) die_nomem();
  if (!stralloc_0(&accept)) die_nomem();

  set_cptarget(accept.s);	/* for copy () */
  set_cpaction(flagconfirm?ACTION_CONFIRM:ACTION_ACCEPT);
  set_cpconfirm(reject.s,quoted.len);
  set_cpwhen(when);

  fdmsg = open_trunc(fnmsg.s);
  if (fdmsg == -1)
    strerr_die2sys(111,FATAL,MSG1(ERR_WRITE,fnmsg.s));
  substdio_fdbuf(&ssmsg,write,fdmsg,msgbuf,sizeof(msgbuf));

  if (qmail_open(&qq) == -1)		/* Open mailer */
    strerr_die2sys(111,FATAL,MSG(ERR_QMAIL_QUEUE));

  hdr_add2s("Mailing-List: ",MSG(TXT_MAILING_LIST));
  if (listid.len > 0)
    hdr_add2("List-ID: ",listid.s,listid.len);
  hdr_datemsgid(when);
  if (flagconfirm)
    hdr_from("-owner");
  else
    hdr_add2s("From: ",reject.s);
  hdr_add2s("Reply-To: ",accept.s);
  if (!flagconfirm && !ismod && flagremote) {	/* if remote admin add -allow- address */
    qmail_puts(&qq,"Cc: ");	/* for ezmlm-gate users */
    strnum[fmt_ulong(strnum,(unsigned long) when)] = 0;
    cookie(hash,key.s,key.len-FLD_ALLOW,strnum,sender,"t");
    if (!stralloc_copy(&line,&outlocal)) die_nomem();
    if (!stralloc_cats(&line,"-allow-tc.")) die_nomem();
    if (!stralloc_cats(&line,strnum)) die_nomem();
    if (!stralloc_append(&line,".")) die_nomem();
    if (!stralloc_catb(&line,hash,COOKIE)) die_nomem();
    if (!stralloc_append(&line,"-")) die_nomem();
    i = str_rchr(sender,'@');
    if (!stralloc_catb(&line,sender,i)) die_nomem();
    if (sender[i]) {
      if (!stralloc_append(&line,"=")) die_nomem();
      if (!stralloc_cats(&line,sender + i + 1)) die_nomem();
    }
    qmail_put(&qq,line.s,line.len);
    qmail_puts(&qq,"@");
    qmail_put(&qq,outhost.s,outhost.len);
    qmail_puts(&qq,"\n");
  }
  qmail_puts(&qq,"To: <");
  if (flagconfirm) {
    if (sender)
      qmail_puts(&qq, sender);
  } else {
    if (!quote(&quoted,&outlocal))
      die_nomem();
    qmail_put(&qq,quoted.s,quoted.len);
    qmail_puts(&qq,"-moderators@");
    qmail_put(&qq,outhost.s,outhost.len);
  }
  qmail_puts(&qq,">\n");
  /* FIXME: Drop the custom subject hack and use hdr_listsubject1 */
  hdr_subject(flagconfirm ? MSG(SUB_CONFIRM_POST) : MSG(SUB_MODERATE));
  if (flagmime) {
    hdr_mime(CTYPE_MULTIPART);
    hdr_boundary(0);
    hdr_ctype(CTYPE_TEXT);
    hdr_transferenc();
  } else
    qmail_puts(&qq,"\n\n");
  copy(&qq,flagconfirm?"text/post-confirm":"text/mod-request",flagcd);
  if (flagcd == 'B') {
    encodeB("",0,&line,2);
    qmail_put(&qq,line.s,line.len);
  }
  if (substdio_put(&ssmsg,returnpath.s,returnpath.len) == -1) die_msg();
  if (substdio_put(&ssmsg,mydtline.s,mydtline.len) == -1) die_msg();
  substdio_fdbuf(&ssin,read,0,inbuf,sizeof(inbuf));

  if (flagmime) {
    hdr_boundary(0);
    hdr_ctype(CTYPE_MESSAGE);
    qmail_puts(&qq, "\n");
  }

  qmail_put(&qq,returnpath.s,returnpath.len);
  qmail_put(&qq,mydtline.s,mydtline.len);
  flaginheader = 1;
  for (;;) {
    if (getln(&ssin,&line,&match,'\n') == -1)
      strerr_die2sys(111,FATAL,MSG(ERR_READ_INPUT));
    if (!match) break;
    if (line.len == 1) flaginheader = 0;
    if (flaginheader) {
      if ((line.len == mydtline.len) &&
		!byte_diff(line.s,line.len,mydtline.s)) {
	close(fdmsg);			/* be nice - clean up */
	unlink(fnmsg.s);
        strerr_die2x(100,FATAL,MSG(ERR_LOOPING));
      }
      if (case_startb(line.s,line.len,"mailing-list:")) {
	close(fdmsg);			/* be nice - clean up */
	unlink(fnmsg.s);
        strerr_die2x(100,FATAL,MSG(ERR_MAILING_LIST));
      }
    }

    if (flagbody || flaginheader)	/* skip body if !flagbody */
      qmail_put(&qq,line.s,line.len);
    if (substdio_put(&ssmsg,line.s,line.len) == -1) die_msg();
  }

  if (flagmime)
    hdr_boundary(1);

/* close archive before qmail. Loss of qmail will result in re-run, and   */
/* worst case this results in a duplicate msg sitting orphaned until it's */
/* cleaned out.                                                           */

  if (substdio_flush(&ssmsg) == -1) die_msg();
  if (fsync(fdmsg) == -1) die_msg();
  if (fchmod(fdmsg,MODE_MOD_MSG | 0700) == -1) die_msg();
  if (close(fdmsg) == -1) die_msg(); /* NFS stupidity */

  close(fdlock);

  if (flagconfirm) {
    qmail_from(&qq,reject.s);			/* envelope sender */
  } else {
    if (!stralloc_copy(&line,&outlocal)) die_nomem();
    if (!stralloc_cats(&line,"-return-@")) die_nomem();
    if (!stralloc_cat(&line,&outhost)) die_nomem();
    if (!stralloc_0(&line)) die_nomem();
    qmail_from(&qq,line.s);			/* envelope sender */
  }
  if (flagconfirm)				/* to sender */
    qmail_to(&qq,sender);
  else if (ismod)				/* to moderator only */
    qmail_to(&qq,mod.s);
  else {
    if (flagself) {				/* to all moderators */
      if (!stralloc_copys(&moderators,"mod")) die_nomem();
      if (!stralloc_0(&moderators)) die_nomem();
    }
    putsubs(moderators.s,0,52,subto);
  }

  if (*(err = qmail_close(&qq)) == '\0') {
      strnum[fmt_ulong(strnum,qmail_qp(&qq))] = 0;
      strerr_die2x(0,"ezmlm-store: info: qp ",strnum);
  } else
      strerr_die4x(111,FATAL,MSG(ERR_TMP_QMAIL_QUEUE),": ",err+1);
}
